imap_processing/ultra/l1b/ultra_l1b_extended.py

```diff
@@ -649,3 +649,111 @@ def get_energy_ssd(de_dataset: xarray.Dataset, ssd: np.ndarray) -> NDArray[np.fl
     energy_norm = get_energy_norm(ssd, composite_energy)
 
     return energy_norm
+
+
+def determine_species_pulse_height(
+    energy: np.ndarray, tof: np.ndarray, r: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Determine the species for pulse-height events.
+
+    Species is determined from the particle energy and velocity.
+    For velocity, the particle TOF is normalized with respect
+    to a fixed distance dmin between the front and back detectors.
+    The normalized TOF is termed the corrected TOF (ctof).
+    Particle species are determined from
+    the energy and ctof using a lookup table.
+
+    Further description is available on pages 42-44 of
+    IMAP-Ultra Flight Software Specification document
+    (7523-9009_Rev_-.pdf).
+
+    Parameters
+    ----------
+    energy : np.ndarray
+        Energy from the SSD event (keV).
+    tof : np.ndarray
+        Time of flight of the SSD event (tenths of a nanosecond).
+    r : np.ndarray
+        Path length (hundredths of a millimeter).
+
+    Returns
+    -------
+    ctof : np.array
+        Corrected TOF.
+    bin : np.array
+        Species bin.
+    """
+    z_dstop = 2.6 / 2  # position of stop foil on Z axis (mm)
+    z_ds = 46.19 - z_dstop  # position of slit on Z axis (mm)
+    df = 3.39  # distance from slit to foil (mm)
```

Comment:
I see these values defined in multiple functions. It seems like you really should have one place where all of the instrument constants are defined. For EMUS I did something like this using a YAML file and wrote an accessor function that would retrieve instrument parameters from that file. It seems like others define things like this in an `__inti__.py` file. I don't know if there is consistency across the project, but defining the same constant in multiple locations should definitely be avoided.